### ISSUE TITLE
Web audio is not restricted while prerendering

### ISSUE THREAD

**Issue by domenic** [WICG/nav-speculation#165]
My read of the current spec at https://wicg.github.io/nav-speculation/prerendering.html is that we have gated audio/video elements, and web speech API, but not any part of web audio.

**Comment by domenic** [comment]
I'm not sure what all the web audio entry points that might play sound are. One might be the `AudioContext` constructor, which currently has a vague, implementation-defined [allowed to start](https://webaudio.github.io/web-audio-api/#allowed-to-start). We could say that prerendering pages are never allowed to start.

**Comment by toyoshim** [comment]
AudioContext can be constructed synchronously, but it doesn't start audio playback if it is constructed without user gesture. In such a case, AudioContext::resume() is needed to start playing. So, we can make the method deferred while prerendering? My assumption is user gestures could not be made while prerendering.

**Comment by toyoshim** [comment]
fyi, @hoch

**Comment by domenic** [comment]
I think that approach sounds reasonable (and indeed, user gestures cannot be made while prerendering). I'm a bit unsure how it fits into the Web Audio spec. https://webaudio.github.io/web-audio-api/ is very hard to read (lots of insertions and deletions for some reason) but it only mentions activation once: https://webaudio.github.io/web-audio-api/#allowed-to-start . It sounds like it's implementation-defined?

Maybe we can just modify https://webaudio.github.io/web-audio-api/#allowed-to-start to say you're never "allowed to start" during prerendering?

**Comment by toyoshim** [comment]
Ah, I see. The user gesture requirement was added later. So, the spec just says it is kind of implementation specific for now even though all browsers are changed to require a user gesture now.
I agreed that our spec patch would say `you're never "allowed to start" during prerendering`.

As far as I know, an explicit call of resume() is needed, and the Web Audio best practice is
 - Construct AudioContext with a user gesture
 - If code was already there and change is difficult, insert AudioContext::resume() at the first user gesture point
 
 resume() may have two choices.
1) reject with InvalidStateError (maybe no patch is needed if we don't allow us to start during prerendering at the AudioContext construction time)
2) [DelayWhilePrerendering]

From the implementation perspective, it just works without any change if we take 1) ?

**Comment by toyoshim** [comment]
Looked at our implementation.
It seems we define AutoplayPolicy, and it can be AutoplayPolicy::kNoUserGestureRequired.
With this case, the AudioContext always starts playback, and users may not expect resume() call. So if prerender disallow it, the page may be broken?
Probably, we say if you are not allowed to start due to prerendering, resume request will be automatically enqueued?

I will check the AutoplayPolicy's details a little.

**Comment by toyoshim** [comment]
OK, Content Settings's Sound can be set to `Allow` to supoport autoplay without user gestures.
Also running with Web App manifest (as a standalone mode?) will permit it.

So, we need some implementation changes to follow up the autoplay permitted case.

**Comment by domenic** [comment]
> Probably, we say if you are not allowed to start due to prerendering, resume request will be automatically enqueued?

This is what it looks like in the spec. In particular I don't think it will reject with "InvalidStateError", I think it will

> append promise to [[[pending promises]]](https://webaudio.github.io/web-audio-api/#dom-baseaudiocontext-pending-promises-slot) and [[[pending resume promises]]](https://webaudio.github.io/web-audio-api/#dom-audiocontext-pending-resume-promises-slot) and abort these steps, returning promise.

which seems like what we want.

For the implementation, I have no knowledge :)

**Comment by toyoshim** [comment]
Ah, sorry. "InvalidStateError" was made for another case, `not fully active` case and it was different from user gesture existence.

pending resume promise is a little tricky. It isn't resolved automatically when the blocking reason, user gesture, is unblocked, but all pending resume promises will be resolved when another resume is called after a user gesture is made.

So the possible minimum spec change is:
to modify https://webaudio.github.io/web-audio-api/#allowed-to-start to say you're never "allowed to start" during prerendering.

If this isn't enough for autoplay enabled sites (they don't call resume at all), we will need yet another automated resume mechanism. For example, if the AudioContext wasn't allowed to start only for prerendering reason, it will automatically resume on activation?

**Comment by domenic** [comment]
Ah, that makes sense. Yes, I agree with your analysis. The minimum change might work, but might lead to sites that would otherwise autoplay being muted until a user gesture. So adding

> if the AudioContext wasn't allowed to start only for prerendering reason, it will automatically resume on activation

would be safest.

**Comment by hoch** [comment]
The discussion so far makes sense to me. Web Audio API also has OfflineAudioContext, but we can treat it like a Worker thread. (I am not sure how Prerendering actually handles a Worker)

So the potential spec change for Web Audio API is:

> never "allowed to start" during prerendering.

And the proposed spec change for this `nav-speculation` work is:

> if the AudioContext wasn't allowed to start only for prerendering reason, it will automatically resume on activation

Did I get that right?

**Comment by toyoshim** [comment]
> @hoch 
Thank you for your feedback here and in the PR.
You may already find answers in the PR, but just in case, let me leave some more context here.

Prerender restricts powerful features that have side effects until users actually trigger the navigation to the prerendered page, and the restriction was realized by deferring promise resolution so that JavaScript can continue to run. Before this change, we cancelled prerendering  on AudioContext uses as the API doesn't have an async mojo entry call that our mojo based general deferral mechanism can handle.

As now we are expecting a potential partner who really wants Web Audio, we start working on this spec work. From the prerendering restriction perspective, audio playback capability would be the only surface that is powerful and has a side effect. So, running OfflineAudioContext sounds fine and safe for prerendering, if I understand the capability correctly.

> Did I get that right?
Yes. If my code reading is correct, autoplay can be permitted via content settings, or in some modes running with Web App Manifests on desktop. So, we need an auto resume for such use cases so that the existing page just works with prerendering. Our goal is to make existing pages just work even if the page doesn't understand how the prerendering works. So, auto-resume would be required. That's my conclusion here.
